backend/search/basex_search.py

```diff
@@ -22,11 +25,63 @@ def check_db_name(db_name: str) -> bool:
     return all(x in ALLOWED_DBNAME_CHARS for x in db_name)
 
 
-def generate_xquery_search(basex_db: str, xpath: str) -> str:
+def check_xquery_variable_name(variable: str) -> bool:
+    if variable[0] != '$':
+        return False
+    return all(x in ALLOWED_VARNAME_CHARS for x in variable[1:])
+
+
+def generate_xquery_for_variables(variables):
+    let_fragment = ''
+    return_fragment_inside = ''
+    if variables is None:
+        variables = []
+    for variable in variables:
+        name = variable['name']
+        path = variable['path']
+        if not check_xquery_variable_name(name):
+            raise ValueError('Invalid XQuery variable: {}'.format(name))
+        if not check_xpath(path):
+            raise ValueError('Invalid XPath: {}'.format(path))
+        properties = variable.get('props', {})
+        # Create fragment for let clause in XQuery for variables
+        if name != '$node':
+            # The root node is already declared in the query itself,
+            # do not declare it again
+            let_fragment += ' let {} := ({})[1]'.format(
+                name, path
+            )
+        # Create fragment for return clause in XQuery for variables
+        properties_xml = ''
+        for prop_name in properties:
+            prop_expression = properties[prop_name]
+            if not all(x in ALLOWED_VARNAME_CHARS for x in prop_name):
+                raise ValueError(
+                    'Property name should be valid XML name: {}'
+                    .format(prop_name)
+                )
+            if '"' in prop_expression or '}' in prop_expression:
+                raise ValueError(
+                    'Property expression cannot contain double quotation '
+                    'mark or closing accolade: {}'.format(prop_expression)
+                )
+            properties_xml += ' ' + prop_name + '="{' + prop_expression + '}"'
+        return_fragment_inside += '<var name="' + name + '"' + \
+            properties_xml + '>{' + name + '/@*}</var>'
+    if return_fragment_inside:
+        return_fragment = '<vars>' + return_fragment_inside + '</vars>'
+    else:
+        return_fragment = ''
+    return let_fragment, return_fragment
+
+
+def generate_xquery_search(basex_db: str, xpath: str, variables=[]) -> str:
```

Comment:
`[]` in a default argument is tricky, because python will allocate a list just once and use it for all function calls. In your case it doesn't matter because you don't edit the list, but it's something to avoid.

(I learned this on a job interview some years ago)